Zend/zend_API.h

```diff
@@ -1834,6 +1849,34 @@ static zend_always_inline int zend_parse_arg_str_or_array_ht(
 	return 1;
 }
 
+static zend_always_inline int zend_parse_arg_str_or_long(zval *arg, zend_string **dest_str, zend_long *dest_long,
+	zend_bool *is_null, int allow_null)
+{
+	if (allow_null) {
+		*is_null = 0;
+	}
+	if (EXPECTED(Z_TYPE_P(arg) == IS_STRING)) {
+		*dest_str = Z_STR_P(arg);
+	} else if (EXPECTED(Z_TYPE_P(arg) == IS_LONG)) {
+		*dest_str = NULL;
+		*dest_long = Z_LVAL_P(arg);
+	} else if (allow_null && EXPECTED(Z_TYPE_P(arg) == IS_NULL)) {
+		*dest_str = NULL;
+		*is_null = 1;
+	} else {
+		if (zend_parse_arg_long_slow(arg, dest_long)) {
+			*dest_str = NULL;
+			return 1;
+		} else if (zend_parse_arg_str_slow(arg, dest_str)) {
+			*dest_long = 0;
+			return 1;
+		} else {
+			return 0;
+		}
```

Comment:
I'd suggest moving this into a `zend_parse_arg_str_or_long_slow` function, to have the whole slow-path out-of-line. You can then also do only one strict types check before passing to the _weak functions.